server/server.go

```diff
@@ -808,12 +855,53 @@ type handlerSwitcher struct {
 	contentTypeToHandler map[string]http.Handler
 }
 
+// Add JSON API paths to NoCsrfHeaderPatterns slice that allow unauthenticated
+// GET requests, so that they will not return X-CSRF-Token header on response.
+var NoCsrfHeaderPattern = []*regexp.Regexp{
+	regexp.MustCompile(`/api/version.*`),
+	regexp.MustCompile(`/api/v1/settings.*`),
+}
+
 func (s *handlerSwitcher) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if urlHandler, ok := s.urlToHandler[r.URL.Path]; ok {
 		urlHandler.ServeHTTP(w, r)
 	} else if contentHandler, ok := s.contentTypeToHandler[r.Header.Get("content-type")]; ok {
 		contentHandler.ServeHTTP(w, r)
 	} else {
+		// If we have a valid authorization header for API access, skip CSRF protection for this request
+		// This is accomplished by setting the key gorilla.csrf.Skip in the request's context.
+		//
+		// We don't validate the JWT token here - this is done at API level. But we make sure that the
+		// request does not carry a cookie named "argocd.token" as sent by the browser. This would
+		// circumvent CSRF protection.
+		//
+		// API access from scripts or other clients therefore MUST always set "Authorization" header
```

Comment:
breaking change?

Reply:
Yes, it breaks write-access (PUT, POST, DELETE) to the API using a Cookie for authorization, and consumers - if having used cookies for API access before - should supply their token in the `Authorization` header instead.

Reply:
I’m not sure we should change this? @jessesuen ?

Reply:
There's not much other option - except having the API client also get a valid CSRF token for every POST request. Either way would break existing (insecure) behaviour, there's really no way I can think of that would change this. :thinking: